### ISSUE TITLE
Requesting large ranges in webdav results in poor performance

### ISSUE THREAD

**Issue by dratini0** [owncloud/core#15321]
I was trying to mount a webdav share with GVFS, and then play a 10G video, hoping it would work, and it didn't. Turns out it requested ranges from the file, which OC apparently does not like.

### Steps to reproduce
1. Put a large file on an Owncloud server
2. Try to request a huge chunk of it using webdav and the HTTP range header. See below for an example.

### Expected behaviour
The download should start immediately

### Actual behaviour
Owncloud went ahead to create a temporary file equal in size to the requested range, which took about a minute (happens for local files and local shares)

### Server configuration
**Operating system**:
Debian 7.8 amd64
**Web server:**
nginx 1.6.2-5~bpo70+1
**Database:**
MySQL 5.5.41-0+wheezy1
**PHP version:**
PHP 5.4.39-0+deb7u2
**ownCloud version:** (see ownCloud admin page)
ownCloud 8.0.2-8
**Updated from an older ownCloud or fresh install:**
Have been rolling since 6.x
**List of activated apps:**
 - activity
 - calendar
 - contacts
 - documents
 - external
 - files
 - files_external
 - files_pdfviewer
 - files_sharing
 - files_texteditor
 - files_trashbin
 - files_versions
 - files_videoviewer
 - gallery
 - news
 - notes

**The content of config/config.php:**

```php
<?php
$CONFIG = array (
  'datadirectory' => '/home/oc_data',
  'dbtype' => 'mysql',
  'version' => '8.0.2.0',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'installed' => true,
  'forcessl' => false,
  'loglevel' => '0',
  'log_authfailip' => true,
  'theme' => '',
  'maintenance' => false,
  'mail_smtpmode' => 'php',
  'overwrite.cli.url' => '/owncloud',
);
```

**Are you using external storage, if yes which one:** local

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** none

### Client configuration
**Browser:**
gvfs 1.22.2, netcat 1.105-7

**Operating system:**
Debian Jessie amd64

A request that jammed my owncloud: (the file is 10704943577 bytes long)
```
GET /owncloud/remote.php/webdav/10G-Movie.mkv HTTP/1.1
Host: ...
Accept-Encoding: gzip, deflate
Accept-Language: en-us, en;q=0.9, en;q=0.8
Connection: Keep-Alive
Range: bytes=40960-
User-Agent: gvfs/1.22.2
Authorization: Basic ...
Cookie: ...
```

P. s.: Sorry, I haven't had the opportunity to try in a newer install, but I haven't seen any discussion regarding the issue.

**Comment by dratini0** [comment]
I have found the culprit, it's in upstream sabredav: [/lib/DAV/CorePlugin.php, line 178](https://github.com/fruux/sabre-dav/blob/master/lib/DAV/CorePlugin.php#L178)

Apparently it implements ranges by creating a temporary stream and blindly copying. I will attempt to submit a pull request there, and we will see how it goes.

**Comment by jospoortvliet** [comment]
#14383 might be helpful for this too.

**Comment by PVince81** [comment]
@dratini0 are you still seeing this in 8.1.4 or 8.2 which contain an updated Sabre version ?

Did I understand well that you are using external storage with the type "local" ? That one should be seekable, so not sure why the seekable check would return false here. The symptoms you see would usually only be observed on non seekable external storages.

**Comment by elvetemedve** [comment]
@PVince81 I've updated Owncloud to version 8.2.1, but `CorePlugin.php` still contains the temporary stream implementation. So the problem is still there.

**Comment by PVince81** [comment]
You mean https://github.com/owncloud/3rdparty/blob/stable8.2/sabre/dav/lib/DAV/CorePlugin.php#L177 ?

Yeah, strange. I'll report this upstream.

**Comment by PVince81** [comment]
Hmm, looks like in Sabre 3 (the next major version) they got rid of the temporary stream https://github.com/fruux/sabre-dav/blob/master/lib/DAV/CorePlugin.php#L175

Here is the ticket for upgrading to Sabre 3: https://github.com/owncloud/core/issues/17039

@evert can you confirm that Sabre 3 will not buffer/copy the whole stream into a temp stream for range requests like Sabre 2 did ?

**Comment by evert** [comment]
It looks like that never made it into the changelog, but I fixed that now.

We did indeed improve this, in this PR:

https://github.com/fruux/sabre-dav/pull/623

Originally appeared in version 2.2.0-alpha4

**Comment by guruz** [comment]
Awesome to see this! :)

**Comment by evert** [comment]
Note that this was 100% by the OP of this ticket, @dratini0

**Comment by PVince81** [comment]
Cool. I see that the fix is available starting with 3.0: https://github.com/fruux/sabre-dav/blob/3.0/lib/DAV/CorePlugin.php#L180

**Comment by DeepDiver1975** [comment]
master is now on sabre 3.0 -> closing
